packages/apollo-links/src/TokenRefreshLink.js

```diff
@@ -0,0 +1,141 @@
+//@flow
+import {
+  ApolloLink,
+  Observable,
+  Operation,
+  NextLink,
+  FetchResult,
+  createOperation,
+} from 'apollo-link';
+import gql from 'graphql-tag';
+import * as R from 'ramda';
+
+import type { TokenRefreshLinkOptions, RefreshTokenResult, RefreshTokenParameters } from './types';
+import { RefreshTokenInvalidError } from './RefreshTokenInvalidError';
+import { hasTokenExpiredError, hasTokenInvalidError } from './utils';
+
+const USER_REFRESH_TOKEN_QUERY = `
+  mutation UserRefreshToken($refreshToken: String!, $email: String!) {
+    userRefreshToken(data: {
+      refreshToken: $refreshToken,
+      email: $email,
+    }) {
+      refreshToken
+      idToken
+    }
+  }
+`;
+
+/**
+ * Token Refresh Link renew authentication token when it's expired.
+ * @param {TokenRefreshLinkOptions} options - The token refresh link options.
+ * @param {Function} options.setRefreshTokenInput - The function which are using for set authentication credentials.
+ * @param {Function} options.authReceived - The callback which called when attempt to refresh authentication is success.
+ * @param {Function} [options.authFailed] - The callback which called when attempt to refresh authentication is failed.
+ */
+class TokenRefreshLink extends ApolloLink {
+  constructor({ setRefreshTokenInput, authReceived, authFailed } : TokenRefreshLinkOptions) {
+    super();
+
+    this.setRefreshTokenInput = setRefreshTokenInput;
+    this.authReceived = authReceived;
```

Comment:
May be we should name it `onAuthReceived`? It isn't clear that `authReceived` is function.
The same comment to `authFailed`.